File: clients/laravel-sdk/src/Outbox/AuditRedaction.php
<?php

declare(strict_types=1);

namespace FlowCatalyst\Outbox;

/**
 * Audit logs must never store passwords or secrets. `operationData` reaches
 * the outbox as the whole command; this walks it and masks anything that
 * looks like a credential before it is serialised into the audit row, so a
 * leak (a command that carries a plaintext token or secret field the way
 * `serviceaccount` Create/Update once did) never reaches `aud_logs`.
 *
 * A key is secret when, lower-cased with `_` and `-` removed, it **ends
 * with** `password`, `passwordhash`, `secret`, `secretref`, `passphrase` or
 * `token`, or **equals** `apikey`, `privatekey`, `authorization` or
 * `cookie`. A secret key's value becomes the string `"***"` — whatever its
 * type — except `null` and booleans, which are kept as-is. `$maskedFields`
 * additionally masks top-level field names the caller declares even though
 * the name rule would keep them (e.g. a config value whose secrecy depends
 * on a sibling field). Objects (associative arrays) and arrays are walked;
 * everything else is untouched. Pure — never mutates `$data`.
 */
final class AuditRedaction
{
    /** @var list<string> */
    private const SECRET_SUFFIXES = [
        'password',
        'passwordhash',
        'secret',
        'secretref',
        'passphrase',
        'token',
    ];

    /** @var list<string> */
    private const SECRET_EXACT = ['apikey', 'privatekey', 'authorization', 'cookie'];

    /**
     * @param array<string, mixed> $data
     * @param list<string> $maskedFields Top-level field names to mask in
     *   addition to the name rule.
     * @return array<string, mixed>
     */
    public static function redact(array $data, array $maskedFields = []): array
    {
        return self::redactObject($data, array_flip($maskedFields));
    }

    /**
     * @param array<string, mixed> $obj
     * @param array<string, int> $topLevelMasked
     * @return array<string, mixed>
     */
    private static function redactObject(array $obj, array $topLevelMasked): array
    {
        $result = [];
        foreach ($obj as $key => $value) {
            if (array_key_exists($key, $topLevelMasked) || self::isSecretKey((string) $key)) {
                $result[$key] = self::maskValue($value);
            } else {
                $result[$key] = self::redactValue($value);
            }
        }

        return $result;
    }

    private static function redactValue(mixed $value): mixed
    {
        // A nested object (a value object, a JsonSerializable) is walked as
        // the document json_encode will write for it, so a secret inside it
        // is caught too.
        if (is_object($value)) {
            $encoded = json_encode($value);
            if ($encoded !== false) {
                $value = json_decode($encoded, true);
            }
        }

        if (is_array($value)) {
            if (array_is_list($value)) {
                return array_map(fn (mixed $item): mixed => self::redactValue($item), $value);
            }

            return self::redactObject($value, []);
        }

        return $value;
    }

    private static function maskValue(mixed $value): mixed
    {
        if ($value === null || is_bool($value)) {
            return $value;
        }

        return '***';
    }

    private static function isSecretKey(string $key): bool
    {
        $normalized = str_replace(['_', '-'], '', strtolower($key));

        if (in_array($normalized, self::SECRET_EXACT, true)) {
            return true;
        }

        foreach (self::SECRET_SUFFIXES as $suffix) {
            if (str_ends_with($normalized, $suffix)) {
                return true;
            }
        }

        return false;
    }
}

File: clients/java-sdk/src/main/java/io/flowcatalyst/sdk/outbox/AuditRedaction.java
package io.flowcatalyst.sdk.outbox;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.util.Iterator;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

/**
 * The audit-redaction rule: an audit row never carries a password, secret
 * or token in the clear, wherever it sits in the command document.
 *
 * <p>The TypeScript, Laravel and Rust SDKs apply the same rule, and every
 * implementation runs the shared {@code audit-redaction-vectors.json} cases.
 * A key is secret when, lower-cased with {@code _} and {@code -} removed, it
 * ends with {@code password}, {@code passwordhash}, {@code secret},
 * {@code secretref}, {@code passphrase} or {@code token}, or equals
 * {@code apikey}, {@code privatekey}, {@code authorization} or
 * {@code cookie}. A secret key's value becomes {@code "***"} whatever its
 * type; {@code null} and booleans are kept. Masked fields are top-level
 * names masked the same way even when the name rule would keep them.
 */
public final class AuditRedaction {

    private static final JsonNodeFactory NODES = JsonNodeFactory.instance;
    private static final String MASK = "***";

    private static final Set<String> SECRET_SUFFIXES = Set.of(
            "password", "passwordhash", "secret", "secretref", "passphrase", "token");

    private static final Set<String> SECRET_EXACT = Set.of(
            "apikey", "privatekey", "authorization", "cookie");

    private AuditRedaction() {}

    /**
     * Redact {@code document}, walking objects and arrays. The input is never
     * mutated: a fresh tree is returned, sharing only unmasked leaf values.
     *
     * @param document the command/operation document ({@code null} passes through)
     * @param masked   top-level field names to mask on top of the name rule
     */
    public static JsonNode redact(JsonNode document, Set<String> masked) {
        if (document == null) return null;
        return redactNode(document, masked == null ? Set.of() : masked, true);
    }

    /** Whether {@code key} is secret under the name rule. */
    public static boolean isSecretKey(String key) {
        String normalized = key.toLowerCase(Locale.ROOT).replace("_", "").replace("-", "");
        if (SECRET_EXACT.contains(normalized)) return true;
        for (String suffix : SECRET_SUFFIXES) {
            if (normalized.endsWith(suffix)) return true;
        }
        return false;
    }

    private static JsonNode redactNode(JsonNode node, Set<String> maskedTopLevel, boolean topLevel) {
        if (node.isObject()) {
            ObjectNode out = NODES.objectNode();
            Iterator<Map.Entry<String, JsonNode>> fields = node.fields();
            while (fields.hasNext()) {
                Map.Entry<String, JsonNode> entry = fields.next();
                String key = entry.getKey();
                JsonNode value = entry.getValue();
                boolean mask = isSecretKey(key) || (topLevel && maskedTopLevel.contains(key));
                out.set(key, mask ? maskValue(value) : redactNode(value, maskedTopLevel, false));
            }
            return out;
        }
        if (node.isArray()) {
            ArrayNode out = NODES.arrayNode(node.size());
            for (JsonNode element : node) {
                out.add(redactNode(element, maskedTopLevel, false));
            }
            return out;
        }
        return node;
    }

    private static JsonNode maskValue(JsonNode value) {
        if (value == null || value.isNull() || value.isBoolean()) return value;
        return NODES.textNode(MASK);
    }
}
